supervisor/homeassistant/websocket.py

```diff
@@ -22,41 +26,97 @@ class WSClient:
     """Home Assistant Websocket client."""
 
     def __init__(
-        self, ha_version: AwesomeVersion, client: aiohttp.ClientWebSocketResponse
+        self,
+        loop: asyncio.BaseEventLoop,
+        ha_version: AwesomeVersion,
+        client: aiohttp.ClientWebSocketResponse,
     ):
         """Initialise the WS client."""
-        self.ha_version: AwesomeVersion = ha_version
-        self.client: aiohttp.ClientWebSocketResponse = client
-        self.message_id: int = 0
-        self._lock: asyncio.Lock = asyncio.Lock()
+        self.ha_version = ha_version
+        self._client = client
+        self._message_id: int = 0
+        self._loop = loop
+        self._futures: dict[int, asyncio.Future[dict]] = {}
+
+    @property
+    def connected(self) -> bool:
+        """Return if we're currently connected."""
+        return self._client is not None and not self._client.closed
+
+    async def close(self) -> None:
+        """Close down the client."""
+        if not self._client.closed:
+            await self._client.close()
+
+    async def async_send_command(self, message: dict[str, Any]) -> dict | None:
+        """Send a websocket message, and return the response."""
+        self._message_id += 1
+        message["id"] = self._message_id
+        self._futures[message["id"]] = self._loop.create_future()
+        _LOGGER.debug("Sending: %s", message)
+        try:
+            await self._client.send_json(message)
+        except ConnectionError as err:
+            raise HomeAssistantWSConnectionError(err) from err
 
-    async def async_send_command(self, message: dict[str, Any]):
-        """Send a websocket command."""
-        async with self._lock:
-            self.message_id += 1
-            message["id"] = self.message_id
+        try:
+            return await self._futures[message["id"]]
+        finally:
+            self._futures.pop(message["id"])
 
-            _LOGGER.debug("Sending: %s", message)
-            try:
-                await self.client.send_json(message)
-            except ConnectionError:
-                return
+    async def start_listener(self) -> None:
+        """Start listening to the websocket."""
+        if not self.connected:
+            raise HomeAssistantWSConnectionError("Not connected when start listening")
 
-            try:
-                response = await self.client.receive_json()
-            except ConnectionError:
-                return
+        try:
+            while self.connected:
+                await self._receive_json()
+        except HomeAssistantWSConnectionError:
+            pass
```

Comment:
.. then we can do something better here:

```
            while self.connected:
                if not await self._receive_json():
                    _LOGGER.debug("Websocket connection closed gracefully.");
                    break
        except HomeAssistantWSConnectionError:
             _LOGGER.warning("Websocket connection closed with connection error.");
```

Reply:
It is mainly that `pass` which I don't like. We could also just add a debug message here, but as stated above, I think its worth differentiate these cases. After all this is going to be a ever more important communication channel, having it sold is worth the effort IMHO :)